### PR TITLE
refactor!: Switch SourceLink to container

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR changes the `SourceLink` back from a base class to a concrete type without inheritance. It now contains a type erased upstream sourcelink (right now implement by `std::any`, but we can change this to use something with small buffer optimization).

Contains #1512 at the moment.

### CODE REVIEW

**Comment by paulgessinger** [comment]
@tboldagh what do you think?

**Comment by paulgessinger** [comment]
Ok, the orthogonal seeding test and the ExaTrkX compilation still need fixes here.

File: Core/include/Acts/SpacePointFormation/detail/SpacePointBuilder.ipp
Comment (paulgessinger) [reply Core/include/Acts/SpacePointFormation/detail/SpacePointBuilder.ipp:14]: I think so, but I'm still fighting with the diff after rebasing.

**Comment by paulgessinger** [comment]
This still carries a slowdown of about 10%. I'm trying to optimize the type erasure mechanism to be more efficient.

**Comment by paulgessinger** [comment]
I made some performance comparisons here on 20 events, 100 runs in each config.
perf_Algorithm_AmbiguityResolutionAlgorithm
perf_Algorithm_DigitizationAlgorithm
perf_Algorithm_FatrasSimulation
perf_Algorithm_IterativeVertexFinder
perf_Algorithm_ParticleSelector
perf_Algorithm_SeedingAlgorithm
perf_Algorithm_SpacePointMaker
perf_Algorithm_TrackFindingAlgorithm
perf_Algorithm_TrackParamsEstimationAlgorithm
perf_Algorithm_TrackSelector
perf_Algorithm_TruthSeedSelector
perf_Reader_EventGenerator

**Comment by paulgessinger** [comment]
I need to unblock this from the `Any` developments. Can we move ahead with this with `std::any`, and accept that the overhead we incur is slightly larger than we would/will until the `Any` type with SBO lands?

Opinions? @asalzburger @andiwand @benjaminhuth ?

**Comment by andiwand** [comment]
makes sense to me as the performance seems to be mostly lost in the ambi resolution which is not optimal anyway

**Comment by benjaminhuth** [comment]
Yeah sounds good. And I think `Acts::Any` will land quite soon, right?

**Comment by paulgessinger** [comment]
There are 100 million edge cases so no ETA on `Acts::Any` yet.

**Comment by paulgessinger** [comment]
Inclusively, a 20 event ODD job goes from about 10.0s to about 11.2s on `-O3`.

Also in this case the largest part seems to indeed be from the Ambiguity resolution. Space point creation and track selection is also slower, but those are generally not very slow to begin with, so it's a smaller effect.

perf_Algorithm_SpacePointMaker
perf_Algorithm_TrackSelector
perf_Algorithm_AmbiguityResolutionAlgorithm

**Comment by paulgessinger** [comment]
I think this is ready for review. @benjaminhuth, @andiwand @asalzburger ?

File: Core/include/Acts/EventData/MultiTrajectory.hpp
Comment (andiwand) [inline Core/include/Acts/EventData/MultiTrajectory.hpp:620]: intended change?

File: Core/include/Acts/EventData/MultiTrajectory.hpp
Comment (andiwand) [inline Core/include/Acts/EventData/MultiTrajectory.hpp:704]: remove dead code?

File: Examples/Framework/include/ActsExamples/EventData/Measurement.hpp
Comment (andiwand) [inline Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:51]: dead code?

File: Examples/Framework/include/ActsExamples/EventData/SimSpacePoint.hpp
Comment (andiwand) [inline Examples/Framework/include/ActsExamples/EventData/SimSpacePoint.hpp:150]: dead code?

**Review by andiwand — APPROVED** [review]
looks good

File: Core/include/Acts/EventData/MultiTrajectory.hpp
Comment (paulgessinger) [reply Core/include/Acts/EventData/MultiTrajectory.hpp:620]: nope, dang

**Comment by paulgessinger** [comment]
@andiwand can you have another quick look and reapprove?

**Review by andiwand — APPROVED** [review]

**Pull request merged**
